File: scala/plugins/Onepassword.scala
// 1Password Connect, as a voxgig/plugin definition.

package com.voxgig.sekreto.plugins

import com.voxgig.sekreto.*
import com.voxgig.sekreto.Providers.checkaddr

/** 1Password, through a Connect server.
  *
  * The item titled `api.token` (titles keep their dots), in the named
  * vault. The value is the field with purpose PASSWORD, or the field
  * labelled `value`. A vault that cannot be found is an error - config
  * names it, so its absence is a broken store, not a missing secret.
  */
class Onepassword(
    addr: Option[String] = None,
    token: Option[String] = None,
    vault: Option[String] = None,
) extends Provider:

  private var vaultid: Option[String] = None

  private def auth(): Map[String, String] =
    Map("authorization" -> s"Bearer ${token.getOrElse("")}")

  private def resolvevault(useaddr: String): String =
    val want = vault.getOrElse("")
    if want.isEmpty then throw SekretoError("sekreto: onepassword: no vault")

    val res = fetchjson("GET", s"$useaddr/v1/vaults", auth())

    val list = res.body.asarr
    if 200 != res.status || list.isEmpty then
      throw SekretoError(s"sekreto: onepassword error: ${res.status}: listing vaults")

    val found = list.get.find: entry =>
      val id = entry.dig("id").text
      id.contains(want) || entry.dig("name").text.contains(want)

    found match
      case Some(entry) => entry.dig("id").text.getOrElse("")
      case None        => throw SekretoError(s"sekreto: onepassword: no vault named $want")

  override def lookup(name: String): Option[String] =
    checkname(name)

    val useaddr = trimslash(addr.getOrElse(""))
    if useaddr.isEmpty then throw SekretoError("sekreto: onepassword: no addr")
    checkaddr(useaddr)

    val id = vaultid.getOrElse:
      val resolved = resolvevault(useaddr)
      vaultid = Some(resolved)
      resolved

    val filter = uriescape(s"""title eq "$name"""")
    val found = fetchjson("GET", s"$useaddr/v1/vaults/$id/items?filter=$filter", auth())

    val items = found.body.asarr
    if 200 != found.status || items.isEmpty then
      throw SekretoError(s"sekreto: onepassword error: ${found.status}: finding $name")

    if items.get.isEmpty then None
    else
      val itemid = items.get.head.dig("id").text.getOrElse("")
      val item = fetchjson("GET", s"$useaddr/v1/vaults/$id/items/$itemid", auth())

      if 200 != item.status then
        throw SekretoError(s"sekreto: onepassword error: ${item.status}: reading $name")

      val fields = item.body.dig("fields").asarr.getOrElse(List.empty)

      fields.find(field => field.dig("purpose").asstr.contains("PASSWORD")) match
        case Some(field) => field.dig("value").text
        case None =>
          fields.find(field => field.dig("label").asstr.contains("value")) match
            case Some(field) => field.dig("value").text
            case None        => None

  override def describe(): String = s"onepassword:${vault.getOrElse("")}"

/** The `onepassword` provider kind, as a voxgig/plugin definition. A consumer
  * imports this and hands it to `Sekreto`; a consumer that does not gets a
  * `Sekreto` that cannot build a chain entry of that kind.
  */
val onepassword: Definition = providerplugin("onepassword", spec =>
  Onepassword(spec.addr, spec.token, spec.vault))

File: kotlin/plugins/Onepassword.kt
// 1Password Connect, as a voxgig/plugin definition.

package com.voxgig.sekreto.plugins

import com.voxgig.sekreto.Definition
import com.voxgig.sekreto.Provider
import com.voxgig.sekreto.Providers.checkaddr
import com.voxgig.sekreto.SekretoError
import com.voxgig.sekreto.checkname
import com.voxgig.sekreto.providerplugin

/**
 * 1Password, through a Connect server.
 *
 * The item titled `api.token` (titles keep their dots), in the named
 * vault. The value is the field with purpose PASSWORD, or the field
 * labelled `value`. A vault that cannot be found is an error - config
 * names it, so its absence is a broken store, not a missing secret.
 */
class Onepassword(
    private val addr: String? = null,
    private val token: String? = null,
    private val vault: String? = null,
) : Provider {

    private var vaultid: String? = null

    private fun auth(): Map<String, String> =
        mapOf("authorization" to "Bearer ${token ?: ""}")

    private fun resolvevault(useaddr: String): String {
        val want = vault ?: ""
        if (want.isEmpty()) {
            throw SekretoError("sekreto: onepassword: no vault")
        }

        val res = fetchjson("GET", "$useaddr/v1/vaults", auth())

        val list = res.body?.asarr
        if (200 != res.status || null == list) {
            throw SekretoError("sekreto: onepassword error: ${res.status}: listing vaults")
        }

        for (entry in list) {
            val id = entry.dig("id")?.text
            if (want == id || want == entry.dig("name")?.text) {
                return id ?: ""
            }
        }

        throw SekretoError("sekreto: onepassword: no vault named $want")
    }

    override fun lookup(name: String): String? {
        checkname(name)

        val useaddr = trimslash(addr ?: "")
        if (useaddr.isEmpty()) {
            throw SekretoError("sekreto: onepassword: no addr")
        }
        checkaddr(useaddr)

        val id = vaultid ?: resolvevault(useaddr).also { vaultid = it }

        val filter = uriescape("title eq \"$name\"")
        val found = fetchjson("GET", "$useaddr/v1/vaults/$id/items?filter=$filter", auth())

        val items = found.body?.asarr
        if (200 != found.status || null == items) {
            throw SekretoError("sekreto: onepassword error: ${found.status}: finding $name")
        }

        if (items.isEmpty()) {
            return null
        }

        val item = fetchjson(
            "GET",
            "$useaddr/v1/vaults/$id/items/${items[0].dig("id")?.text}",
            auth(),
        )

        if (200 != item.status) {
            throw SekretoError("sekreto: onepassword error: ${item.status}: reading $name")
        }

        val fields = item.body?.dig("fields")?.asarr ?: emptyList()

        for (field in fields) {
            if ("PASSWORD" == field.dig("purpose")?.asstr) {
                return field.dig("value")?.text
            }
        }
        for (field in fields) {
            if ("value" == field.dig("label")?.asstr) {
                return field.dig("value")?.text
            }
        }

        return null
    }

    override fun describe(): String = "onepassword:${vault ?: ""}"
}

/** The `onepassword` provider kind, as a voxgig/plugin definition. A
 * consumer imports this and hands it to `Sekreto`; a consumer that does
 * not is a `Sekreto` that cannot build a `onepassword` chain entry. */
val onepassword: Definition = providerplugin("onepassword") { spec ->
    Onepassword(spec.addr, spec.token, spec.vault)
}
